netmaster/objApi/apiController.go

```diff
@@ -0,0 +1,639 @@
+	}
+
+	// Setup links
+	modeldb.AddLink(&service.Links.App, app)
+	modeldb.AddLinkSet(&app.LinkSets.Services, service)
+
+	// Save the app too since we added the links
+	err := app.Write()
+	if err != nil {
+		return err
+	}
+
+	// Check if user specified any networks
+	if len(service.Networks) == 0 {
+		service.Networks = append(service.Networks, "privateNet")
+	}
+
+	// link service with network
+	for _, netName := range service.Networks {
+		netKey := service.TenantName + ":" + netName
+		network := contivModel.FindNetwork(netKey)
+		if network == nil {
+			log.Errorf("Service: %s could not find network %s", service.Key, netKey)
+			return errors.New("Network not found")
+		}
+
+		// Link the network
+		modeldb.AddLinkSet(&service.LinkSets.Networks, network)
+		modeldb.AddLinkSet(&network.LinkSets.Services, service)
+
+		// save the network
+		err := network.Write()
+		if err != nil {
+			return err
+		}
+	}
+
+	// Check if user specified any endpoint group for the service
+	if len(service.EndpointGroups) == 0 {
+		// Create one default endpointGroup per network
+		for _, netName := range service.Networks {
+			// params for default endpoint group
+			dfltEpgName := service.AppName + "." + service.ServiceName + "." + netName
+			endpointGroup := contivModel.EndpointGroup{
+				Key:         service.TenantName + ":" + dfltEpgName,
+				TenantName:  service.TenantName,
+				NetworkName: netName,
+				GroupName:   dfltEpgName,
+			}
+
+			// Create default endpoint group for the service
+			err = contivModel.CreateEndpointGroup(&endpointGroup)
+			if err != nil {
+				log.Errorf("Error creating endpoint group: %+v, Err: %v", endpointGroup, err)
+				return err
+			}
+
+			// Add the endpoint group to the list
+			service.EndpointGroups = append(service.EndpointGroups, dfltEpgName)
+		}
+	}
+
+	// Link the service and endpoint group
+	for _, epgName := range service.EndpointGroups {
+		endpointGroup := contivModel.FindEndpointGroup(service.TenantName + ":" + epgName)
+		if endpointGroup == nil {
+			log.Errorf("Error: could not find endpoint group: %s", epgName)
+			return errors.New("could not find endpointGroup")
+		}
+
+		// setup links
+		modeldb.AddLinkSet(&service.LinkSets.EndpointGroups, endpointGroup)
+		modeldb.AddLinkSet(&endpointGroup.LinkSets.Services, service)
+
+		// save the endpointGroup
+		err = endpointGroup.Write()
+		if err != nil {
+			return err
+		}
+	}
+
+	// Check if user specified any volume profile
+	if service.VolumeProfile == "" {
+		service.VolumeProfile = "default"
+	}
+
+	volProfKey := service.TenantName + ":" + service.VolumeProfile
+	volProfile := contivModel.FindVolumeProfile(volProfKey)
+	if volProfile == nil {
+		log.Errorf("Could not find the volume profile: %s", service.VolumeProfile)
+		return errors.New("VolumeProfile not found")
+	}
+
+	// fixup default values
+	if service.Scale == 0 {
+		service.Scale = 1
+	}
+
+	// Create service instances
+	for idx := int64(0); idx < service.Scale; idx++ {
```

Comment:
We may need to take care of allocating network resource per instance as well like an IP per instance for a service behind a HAProxy?